pennylane/beta/_qubit_device.py

```diff
@@ -0,0 +1,634 @@
+        can be used by devices in :meth:`apply` for parametric compilation."""
+
+    @classmethod
+    def capabilities(cls):
+
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            model="qubit",
+            supports_finite_shots=True,
+            supports_tensor_observables=True,
+            returns_probs=True,
+        )
+        return capabilities
+
+    def reset(self):
+        """Reset the backend state.
+
+        After the reset, the backend should be as if it was just constructed.
+        Most importantly the quantum state is reset to its initial value.
+        """
+        self._samples = None
+        self._circuit_hash = None
+
+    def execute(self, circuit, **kwargs):
+        """Execute a queue of quantum operations on the device and then
+        measure the given observables.
+
+        For plugin developers: instead of overwriting this, consider
+        implementing a suitable subset of
+
+        * :meth:`apply`
+
+        * :meth:`~.generate_samples`
+
+        * :meth:`~.probability`
+
+        Additional keyword arguments may be passed to the this method
+        that can be utilised by :meth:`apply`. An example would be passing
+        the ``QNode`` hash that can be used later for parametric compilation.
+
+        Args:
+            circuit (~.CircuitGraph): circuit to execute on the device
+
+        Raises:
+            QuantumFunctionError: if the value of :attr:`~.Observable.return_type` is not supported
+
+        Returns:
+            array[float]: measured value(s)
+        """
+        self.check_validity(circuit.operations, circuit.observables)
+
+        self._circuit_hash = circuit.hash
+
+        if circuit.is_batched:
+            # redirect entire execution to other method
+            return self.batch_execute(circuit, **kwargs)
+
+        else:
+            return self.single_execute(circuit, **kwargs)
+
+    def single_execute(self, circuit, **kwargs):
+        """Execute a single circuit."""
+        # apply all circuit operations
+        self.apply(circuit.operations, rotations=circuit.diagonalizing_gates, **kwargs)
+
+        # generate computational basis samples
+        if (not self.analytic) or circuit.is_sampled:
+            self._samples = self.generate_samples()
+
+        # compute the required statistics
+        results = self.statistics(circuit.observables)
+
+        # Ensures that a combination with sample does not put
+        # expvals and vars in superfluous arrays
+        all_sampled = all(obs.return_type is Sample for obs in circuit.observables)
+        if circuit.is_sampled and not all_sampled:
+            return self._asarray(results, dtype="object")
+
+        return self._asarray(results)
+
+    def batch_execute(self, circuit, **kwargs):
+        """Execute a circuit that contains batch tapes."""
+
+        if hasattr(self, "batch_apply"):
+            res = None  # Todo: determine how we can run the qubit device in batch
```

Comment:
Dummy for now. This one is still a bit tricky: can the entire logic of sampling etc be done in batches without loosing the advantages of batching? 
Or do we need another class entirely for HPC/batch-supporting devices? @antalszava would be curious what you think?

Reply:
For now I am just using `QubitDevice` because it has the most updated interface. These changes should probably all be made in `Device`, ultimately...

Reply:
Think it depends: `QubitDevice` after all provides more of scaffolding for devices. If `batch_execute` would be a completely stand-alone method, I'd think it should be easily feasible.

If `batch_execute` would depend on `execute`, then we'd have to consider what data is stored after each execution. As an example, some parts of the logic in `execute` can be tailored such that samples are generated based on the internally stored state for state vector simulators (e.g. `generate_samples` depends on the abstract `analytic_probability` method which for example in `default.qubit` uses `dev.state`). Would we then overwrite the state & samples generated after each execution or store all of them?

Think that one of the crucial questions is to determine if all sample sets would be stored. In the case of state vector simulators, the question would be if they would store their states.

Another thought would be to see how a specific simulator with a `batch_execute` method would look like. We could then see how the main `QubitDevice` API could be adjusted.